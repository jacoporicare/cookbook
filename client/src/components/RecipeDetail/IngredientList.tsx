import React, { Component, ChangeEvent } from 'react';

import { Ingredient } from '../../types';

type Props = {
  ingredients?: Ingredient[];
  servingCount?: number;
};

type State = {
  servingCount: string;
};

class IngredientList extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      servingCount: props.servingCount ? props.servingCount.toString() : '',
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    this.setState({
      servingCount: nextProps.servingCount ? nextProps.servingCount.toString() : '',
    });
  }

  getAmount(amount?: number) {
    if (!amount) {
      return '';
    }

    if (!this.state.servingCount || !this.props.servingCount) {
      return amount.toLocaleString('cs');
    }

    return (
      (amount / this.props.servingCount) *
      Number.parseInt(this.state.servingCount, 10)
    ).toLocaleString('cs');
  }

  handleServingCountChange = (event: ChangeEvent<HTMLInputElement>) => {
    const servingCount = Number.parseInt(event.target.value, 10);

    this.setState({
      servingCount: !Number.isNaN(servingCount) && servingCount > 0 ? servingCount.toString() : '',
    });
  };

  render() {
    const { ingredients, servingCount: initialServingCount } = this.props;
    const { servingCount } = this.state;

    if (!ingredients || !ingredients.length) {
      return <div className="alert alert-info">Žádné ingredience.</div>;
    }

    return (
      <>
        {!!initialServingCount && (
          <div className="form-group">
            <div className="input-group">
              <div className="input-group-addon">Počet porcí</div>
              <input
                type="number"
                value={servingCount}
                onChange={this.handleServingCountChange}
                min={1}
                className="form-control"
              />
            </div>
          </div>
        )}

        <ul className="list-group cb-ingredient-list">
          {ingredients.map(ingredient => {
            const { _id: id, isGroup, name, amount, amountUnit } = ingredient;

            let className = 'list-group-item';
            if (isGroup) {
              className += ' list-group-item-warning';
            }

            return (
              <li key={id} className={className}>
                {isGroup ? (
                  <b>{name}</b>
                ) : (
                  <div className="row">
                    <div className="col-xs-3 text-right">
                      {(amount || amountUnit) && (
                        <b>
                          {this.getAmount(amount)}&nbsp;{amountUnit}
                        </b>
                      )}
                    </div>
                    <div className="col-xs-9">{name}</div>
                  </div>
                )}
              </li>
            );
          })}
        </ul>
      </>
    );
  }
}

export default IngredientList;
