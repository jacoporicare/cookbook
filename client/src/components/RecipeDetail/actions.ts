import { Dispatch } from 'redux';

import { RecipeDetail, StoreState } from '../../types';
import api, { handleError } from '../../api';

export type RecipeDetailAction =
  | { type: 'RECIPE.FETCH.REQUEST' }
  | { type: 'RECIPE.FETCH.SUCCESS'; payload: { recipe: RecipeDetail; slug: string } }
  | { type: 'RECIPE.FETCH.FAILURE' };

export const fetchRecipeRequest = (): RecipeDetailAction => ({
  type: 'RECIPE.FETCH.REQUEST',
});

export const fetchRecipeSuccess = (recipe: RecipeDetail, slug: string): RecipeDetailAction => ({
  type: 'RECIPE.FETCH.SUCCESS',
  payload: {
    recipe,
    slug,
  },
});

export const fetchRecipeFailure = (): RecipeDetailAction => ({
  type: 'RECIPE.FETCH.FAILURE',
});

export function fetchRecipe(slug: string) {
  return (dispatch: Dispatch<StoreState>, getState: () => StoreState) => {
    dispatch(fetchRecipeRequest());

    return api(getState)
      .get<RecipeDetail>(`/api/recipes/${slug}`)
      .then(({ data }) => dispatch(fetchRecipeSuccess(data, slug)))
      .catch(error => {
        handleError(error);
        return dispatch(fetchRecipeFailure());
      });
  };
}
