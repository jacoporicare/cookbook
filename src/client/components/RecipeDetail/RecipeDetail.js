import React from 'react';
import PropTypes from 'prop-types';
import RichText from '../RichText/RichText';
import Ingredients from './IngredientList';

const RecipeDetail = ({ ingredients, servingCount, directions }) =>
  <div className="cb-recipe-detail">
    <div className="row">
      <div className="col-md-3 col-sm-4">
        <h3>Ingredience</h3>
        <Ingredients ingredients={ingredients} servingCount={servingCount} />
      </div>

      <div className="col-md-9 col-sm-8">
        <h3>Postup</h3>
        {directions
          ? <RichText text={directions} />
          : <div className="alert alert-info">Žádný postup.</div>}
      </div>
    </div>
  </div>;

RecipeDetail.propTypes = {
  ingredients: PropTypes.array,
  servingCount: PropTypes.number,
  directions: PropTypes.string,
};

export default RecipeDetail;
