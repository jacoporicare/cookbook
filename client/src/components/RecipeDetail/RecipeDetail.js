import React from 'react';
import PropTypes from 'prop-types';
import RichText from '../RichText/RichText';
import IngredientList from './IngredientList';

const RecipeDetail = ({ ingredients, servingCount, directions }) => (
  <div className="cb-recipe-detail">
    <div className="row">
      <div className="col-md-3 col-sm-4">
        <IngredientList ingredients={ingredients} servingCount={servingCount} />
      </div>

      <div className="col-md-9 col-sm-8">
        {directions ? (
          <RichText text={directions} />
        ) : (
          <div className="alert alert-info">Žádný postup.</div>
        )}
      </div>
    </div>
  </div>
);

RecipeDetail.propTypes = {
  ingredients: PropTypes.array,
  servingCount: PropTypes.number,
  directions: PropTypes.string,
};

export default RecipeDetail;
