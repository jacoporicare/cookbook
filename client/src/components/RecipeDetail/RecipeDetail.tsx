import React from 'react';
import styled from 'react-emotion';

import { Ingredient } from '../../types';
import RichText from '../RichText/RichText';
import IngredientList from './IngredientList';

type Props = {
  ingredients?: Ingredient[];
  servingCount?: number;
  directions?: string;
  lastModifiedDate: string;
  imageUrl?: string;
  imageFullUrl?: string;
};

const Info = styled.p`
  margin-bottom: 20px;
`;

const ImageLink = styled.a`
  float: right;
  position: relative;
  z-index: 2;
`;

const Image = styled.img`
  width: 200px;
  height: 200px;
  margin-left: 15px;
  margin-bottom: 15px;
  border-radius: 4px;
`;

const ImageXs = styled.img`
  width: 100%;
  border-radius: 4px;
  margin-top: 15px;
`;

export default function RecipeDetail({
  ingredients,
  servingCount,
  directions,
  lastModifiedDate,
  imageUrl,
  imageFullUrl,
}: Props) {
  return (
    <>
      <div className="row">
        <div className="col-md-3 col-sm-4">
          <IngredientList ingredients={ingredients} servingCount={servingCount} />
          <Info>{new Date(lastModifiedDate).toLocaleDateString('cs')}</Info>
        </div>

        <div className="col-md-9 col-sm-8">
          {imageUrl && (
            <ImageLink href={imageFullUrl} target="_blank" className="hidden-xs">
              <Image src={imageUrl} />
            </ImageLink>
          )}
          {directions ? (
            <RichText text={directions} />
          ) : (
            <div className="alert alert-info">Žádný postup.</div>
          )}
        </div>
      </div>
      {imageUrl && (
        <a href={imageFullUrl} target="_blank" className="visible-xs">
          <ImageXs src={imageUrl} />
        </a>
      )}
    </>
  );
}
