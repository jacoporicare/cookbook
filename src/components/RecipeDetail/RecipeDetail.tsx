import React from 'react';
import styled from 'react-emotion';

import { Ingredient } from '../../types';
import { colors } from '../../styles/colors';
import { Box, Text, BoxAside, BoxSection, BoxArticle } from '../core';
import { InfoAlert } from '../elements';
import RichText from '../RichText/RichText';
import IngredientList from './IngredientList';

type Props = {
  title: string;
  ingredients?: Ingredient[];
  servingCount?: number;
  directions?: string;
  lastModifiedDate: number;
  imageUrl?: string;
  imageFullUrl?: string;
  userName: string;
};

const ImageBox = styled(Box)`
  float: right;
  position: relative;
  z-index: 2;
`;

const Image = styled('img')`
  width: 200px;
  height: 200px;
  margin-left: 15px;
  margin-bottom: 15px;
  border-radius: 4px;
`;

const ImageXs = styled('img')`
  width: 100%;
  border-radius: 4px;
  margin-top: 15px;
`;

function RecipeDetail({
  title,
  ingredients,
  servingCount,
  directions,
  lastModifiedDate,
  imageUrl,
  imageFullUrl,
  userName,
}: Props) {
  return (
    <>
      <Box display={['block', 'flex']}>
        <BoxAside flex={1} pr={[0, 3]}>
          <IngredientList ingredients={ingredients} servingCount={servingCount} />
          <Box my={4}>
            <Text color={colors.gray600}>Autor:</Text>
            <Box>{userName}</Box>
            <Text color={colors.gray600}>Naposledy upraveno:</Text>
            <Box>{new Date(lastModifiedDate).toLocaleDateString('cs')}</Box>
          </Box>
        </BoxAside>

        <Box flex={3}>
          {imageUrl && (
            <ImageBox display={['none', 'block']}>
              <a href={imageFullUrl} target="_blank">
                <Image src={imageUrl} alt={title} />
              </a>
            </ImageBox>
          )}
          {directions ? (
            <RichText text={directions} />
          ) : (
            <Box mr={[0, !directions && imageUrl ? '215px' : 0]}>
              <InfoAlert>Žádný postup.</InfoAlert>
            </Box>
          )}
        </Box>
      </Box>
      {imageUrl && (
        <Box display={['block', 'none']}>
          <a href={imageFullUrl} target="_blank">
            <ImageXs src={imageUrl} />
          </a>
        </Box>
      )}
    </>
  );
}

export default RecipeDetail;
