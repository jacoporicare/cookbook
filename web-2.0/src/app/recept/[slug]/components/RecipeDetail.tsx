import { Alert, Box, Grid, Paper, Typography } from '@mui/material';
import Image from 'next/image';
import ReactMarkdown from 'react-markdown';

import IngredientList from './IngredientList';
import RecipeHeader from './RecipeHeader';

import RichText from '@/app/components/RichText';
import { INSTANT_POT_TAG } from '@/const';
import { FragmentType, graphql, useFragment } from '@/gql';

export const RecipeDetailFragment = graphql(`
  fragment RecipeDetailItem on Recipe {
    id
    slug
    title
    tags
    directions
    imageUrl(format: WEBP)
    sideDish
    servingCount
    tags
    preparationTime
    imageUrl: imageUrl(format: WEBP)
    imageThumbUrl: imageUrl(size: { width: 800, height: 800 }, format: WEBP)
    lastModifiedDate
    user {
      id
      displayName
    }
    ingredients {
      ...RecipeDetailIngredientItem
    }
  }
`);

export const RecipeDetailUserFragment = graphql(`
  fragment RecipeDetailUser on User {
    id
    isAdmin
  }
`);

type Props = {
  recipe: FragmentType<typeof RecipeDetailFragment>;
  user: FragmentType<typeof RecipeDetailUserFragment>;
};

function RecipeDetail(props: Props) {
  const recipe = useFragment(RecipeDetailFragment, props.recipe);
  const user = useFragment(RecipeDetailUserFragment, props.user);

  return (
    <>
      <RecipeHeader
        isAuthor={Boolean(user && (user.isAdmin || user.id === recipe.user.id))}
        preparationTime={recipe.preparationTime ?? undefined}
        sideDish={recipe.sideDish ?? undefined}
        slug={recipe.slug}
        tags={recipe.tags}
        title={recipe.title}
        onDeleteShow={() => {
          throw new Error('Function not implemented.');
        }}
      />

      <Grid spacing={4} container>
        <Grid lg={3} md={4} xs={12} item>
          <IngredientList
            ingredients={recipe.ingredients}
            servingCount={recipe.servingCount ?? undefined}
          />
        </Grid>

        <Grid lg={7} md={6} xs={12} item>
          <Typography component="h3" variant="h5" gutterBottom>
            Postup
          </Typography>
          {recipe.tags.includes(INSTANT_POT_TAG) && (
            <Paper>
              <Box mb={3}>
                <Alert severity="info">
                  <strong>Instant Pot</strong>
                  <br />
                  Tento recept je určený pro multifunkční hrnec Instant Pot nebo jeho kopie, např.
                  česká Tesla EliteCook K70.
                </Alert>
              </Box>
            </Paper>
          )}
          <Paper>
            {recipe.directions ? (
              <Box p={3}>
                <RichText>
                  <ReactMarkdown>{recipe.directions}</ReactMarkdown>
                </RichText>
              </Box>
            ) : (
              <Alert severity="info">Žádný postup.</Alert>
            )}
          </Paper>
        </Grid>
        <Grid md={2} xs={12} item>
          <Box pt={{ xs: 0, md: '2.525rem' }}>
            {recipe.imageUrl && recipe.imageThumbUrl && (
              <Paper>
                <Box
                  component="a"
                  href={recipe.imageUrl}
                  sx={{
                    display: 'block',
                    position: 'relative',
                    paddingTop: '75%',
                  }}
                >
                  <Image
                    alt={recipe.title}
                    quality={100}
                    sizes="(max-width: 900px) 100vw, 16vw"
                    src={recipe.imageThumbUrl}
                    style={{
                      objectFit: 'cover',
                      borderRadius: '4px',
                    }}
                    fill
                  />
                </Box>
              </Paper>
            )}
            <Box my={2}>
              <Typography color="textSecondary" variant="subtitle1">
                Autor:
              </Typography>
              <Typography variant="body1">{recipe.user.displayName}</Typography>
              <Typography color="textSecondary" variant="subtitle1">
                Naposledy upraveno:
              </Typography>
              <Typography variant="body1">
                {new Date(recipe.lastModifiedDate).toLocaleDateString('cs')}
              </Typography>
            </Box>
          </Box>
        </Grid>
      </Grid>
    </>
  );
}

export default RecipeDetail;
