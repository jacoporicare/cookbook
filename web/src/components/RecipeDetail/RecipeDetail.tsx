import { Alert, Box, Grid, Paper, Typography } from '@mui/material';
import { MouseEvent, useState } from 'react';
import Lightbox from 'react-image-lightbox';

import { Ingredient } from '../../generated/graphql';
import RichText from '../RichText/RichText';

import IngredientList from './IngredientList';

type Props = {
  title: string;
  ingredients?: Omit<Ingredient, '_id'>[];
  servingCount?: number;
  directions?: string;
  lastModifiedDate: string;
  imageUrl?: string;
  imageFullUrl?: string;
  userName: string;
};

function RecipeDetail({
  title,
  ingredients,
  servingCount,
  directions,
  lastModifiedDate,
  imageUrl,
  imageFullUrl,
  userName,
}: Props) {
  const [isImageOpen, setIsImageOpen] = useState(false);

  function handleImageClick(e: MouseEvent) {
    e.preventDefault();
    setIsImageOpen(true);
  }

  return (
    <>
      <Grid spacing={4} container>
        <Grid lg={3} md={4} xs={12} item>
          <IngredientList ingredients={ingredients} servingCount={servingCount} />
        </Grid>

        <Grid lg={7} md={6} xs={12} item>
          <Typography component="h3" variant="h5" gutterBottom>
            Postup
          </Typography>
          <Paper>
            {directions ? (
              <Box p={3}>
                <RichText text={directions} />
              </Box>
            ) : (
              <Alert severity="info">Žádný postup.</Alert>
            )}
          </Paper>
        </Grid>
        <Grid md={2} xs={12} item>
          <Box pt={{ xs: 0, md: '2.525rem' }}>
            {imageUrl && (
              <Paper>
                <Box
                  component="a"
                  href={imageFullUrl}
                  sx={{
                    display: 'block',
                    lineHeight: 0,
                  }}
                  onClick={handleImageClick}
                >
                  <Box
                    alt={title}
                    component="img"
                    src={imageUrl}
                    sx={{
                      width: '100%',
                      borderRadius: '4px',
                    }}
                  />
                </Box>
              </Paper>
            )}
            <Box my={2}>
              <Typography color="textSecondary" variant="subtitle1">
                Autor:
              </Typography>
              <Typography variant="body1">{userName}</Typography>
              <Typography color="textSecondary" variant="subtitle1">
                Naposledy upraveno:
              </Typography>
              <Typography variant="body1">
                {new Date(lastModifiedDate).toLocaleDateString('cs')}
              </Typography>
            </Box>
          </Box>
        </Grid>
      </Grid>
      {isImageOpen && (
        <Lightbox mainSrc={imageFullUrl!} onCloseRequest={() => setIsImageOpen(false)} />
      )}
    </>
  );
}

export default RecipeDetail;
