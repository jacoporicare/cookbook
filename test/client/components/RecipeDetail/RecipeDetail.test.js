import React from 'react';
import renderer from 'react-test-renderer';
import RecipeDetail from '../../../../src/client/components/RecipeDetail/RecipeDetail';

test('<RecipeDetail /> renders correctly', () => {
  const component = renderer.create(
    <RecipeDetail recipe={{}} />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
