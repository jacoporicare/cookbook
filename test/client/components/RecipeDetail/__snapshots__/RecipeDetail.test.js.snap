exports[`test <RecipeDetail /> renders correctly 1`] = `
<div
  className="cb-recipe-detail">
  <div
    className="row">
    <div
      className="col-md-3 col-sm-4">
      <h3>
        Ingredience
      </h3>
      <div
        className="alert alert-info">
        Žádné ingredience.
      </div>
    </div>
    <div
      className="col-md-9 col-sm-8">
      <h3>
        Postup
      </h3>
      <div
        className="alert alert-info">
        Žádný postup.
      </div>
    </div>
  </div>
</div>
`;
